src/particle_filter.cpp

```diff
@@ -1,33 +1,95 @@
 
 // ----------------------------------------------------------------------------------------------------
 
-ParticleFilter::ParticleFilter() : i_current_(0)
+ParticleFilter::ParticleFilter() : min_samples_(0), max_samples_(0), kld_err_(0), kld_z_(0), alpha_slow_(0),
+    alpha_fast_(0), w_slow_(0), w_fast_(0), i_current_(0), kd_tree_(nullptr)
 {
 }
 
 // ----------------------------------------------------------------------------------------------------
 
+void ParticleFilter::configure(tue::Configuration config)
+{
+    // Extra variable needed as tue::Configuration doesn't support
+    // unisgned interters
+    int min, max;
+    config.value("min_particles", min);
+    config.value("max_particles", max);
+    min_samples_ = min;
+    max_samples_ = max;
+
+    kld_err_ = 0.01;
+    kld_z_ = 0.99;
+    config.value("kld_err", kld_err_, tue::config::OPTIONAL);
+    config.value("kld_z", kld_z_, tue::config::OPTIONAL);
+
+    alpha_slow_ = 0;
+    alpha_fast_ = 0;
+    config.value("recovery_alpha_slow", alpha_slow_, tue::config::OPTIONAL);
+    config.value("recovery_alpha_fast", alpha_fast_, tue::config::OPTIONAL);
+
+    std::array<double, 3> cell_size = {0.5, 0.5, 10*M_PI/180};
+    config.value("cell_size_x", cell_size[0], tue::config::OPTIONAL);
+    config.value("cell_size_y", cell_size[1], tue::config::OPTIONAL);
+    config.value("cell_size_theta", cell_size[2], tue::config::OPTIONAL);
+
+    samples_[0].reserve(max_samples_);
+    samples_[1].reserve(max_samples_);
+
+    cluster_cache_.reserve(max_samples_);
+
+    limit_cache_.clear();
+    limit_cache_.resize(max_samples_, 0);
+
+    kd_tree_.reset(new KDTree(max_samples_, cell_size));
+
+    ROS_INFO_STREAM("[ED Localization] min_samples: " << min_samples_ << ", max_samples: " << max_samples_);
+    ROS_INFO_STREAM("[ED Localization] kld_err: " << kld_err_ << ", kld_z: " << kld_z_);
+    ROS_INFO_STREAM("[ED Localization] recovery_alpha_slow: " << alpha_slow_ << ", recovery_alpha_fast: " << alpha_fast_);
+    ROS_INFO_STREAM("[ED Localization] cell_size_x: " << cell_size[0] << ", cell_size_y: " << cell_size[1] << ", cell_size_theta: " << cell_size[2]);
+}
+
+// ----------------------------------------------------------------------------------------------------
+
 ParticleFilter::~ParticleFilter()
 {
 }
 
 // ----------------------------------------------------------------------------------------------------
 
-void ParticleFilter::initUniform(const geo::Vec2& min, const geo::Vec2& max, double t_step,
-                                 double a_min, double a_max, double a_step)
+void ParticleFilter::initUniform(const geo::Vec2& min, const geo::Vec2& max, double a_min, double a_max)
 {
+    clearCache();
+
     std::vector<Sample>& smpls = samples();
 
     smpls.clear();
-    for(double x = min.x; x < max.x; x += t_step)
-        for(double y = min.y; y < max.y; y += t_step)
-            for(double a = a_min; a < a_max; a += a_step)
+
+    const double range_x = max.x - min.x;
```

Comment:
Is there a specific reason why you use const double? I guess it's actually good practice but hardly ever see this elsewhere

Reply:
No, just good practice ;)